### PR TITLE
fix: [2.38-DHIS2-TECH842.2] - Grid Caching is not properly used in analytics

### PR DESCRIPTION
**Pull request by d-bernat**
/api/analytics/dataValueSet.json?dimension=dx:SqD0kovH0CS&dimension=ou:m5NxCAOnDg6&dimension=pe:20210902 where dx is a program indicator returns null value, when the request is issued second time and more within AnalyticsCache TTL
(typically 60sec)

first response:
``` json
{
    "dataValues": [
        {
            "dataElement": "SqD0kovH0CS",
            "period": "20210902",
            "orgUnit": "m5NxCAOnDg6",
            "value": "2.0",
            "storedBy": "[aggregated]",
            "created": "2021-11-23",
            "lastUpdated": "2021-11-23",
            "comment": "[aggregated]"
        }
    ]
}
```
second one:
``` json
{
    "dataValues": [
        {
            "dataElement": "SqD0kovH0CS",
            "period": "20210902",
            "orgUnit": "m5NxCAOnDg6",
            "value": "null",
            "storedBy": "[aggregated]",
            "created": "2021-11-23",
            "lastUpdated": "2021-11-23",
            "comment": "[aggregated]"
        }
    ]
}
```
The problem was identified in the caching of analytics Grid object:

this is object coming to cache:
[dy, ou, pe, value]
[SqD0kovH0CS, m5NxCAOnDg6, 20210902, 2.0]

first get:
[dy, ou, pe, value]
[SqD0kovH0CS, m5NxCAOnDg6, 20210902, null, null, 2.0]

second one:
[dy, ou, pe, value]
[SqD0kovH0CS, m5NxCAOnDg6, 20210902, null, null, null, null, 2.0]

We are caching the grid and modifying it later within one single request. Because this is in memory cache inside the runtime JVM, only reference is cached, there is no serialization on the way. After the caching, cached object is modified (add two rows for coc and aoc). Within the next request we modified already modified object so the grid is not readable correctly anymore until the cache expiration.

Solution description:
in the class DataAgregator there is critical method getAggregatedDataValueGrid. It is kind of factory. In many steps the grid object is created and filled up with data. In the mid of this process the incomplete Grid object is cached. In the next steps is this object modified (cache has just reference on grid) and finally returned. To avoid the caching of incomplete grid object the "get" operation is done before the factory pipeline and "put" one is done after it.

### CODE REVIEW

**Review by gnespolino — COMMENTED** [review]
So if I'm not wrong we will now cache a reference to the grid itself rather than its copy.
Will it work when using redis cache (which to my knowledge will involve serialization/deserialization) ?
AFAIK in the grid there are still references to hibernate proxies in this way and I'm not sure it will work when using a cache which involves serialization.

**Comment by d-bernat** [comment]
> So if I'm not wrong we will now cache a reference to the grid itself rather than its copy. Will it work when using redis cache (which to my knowledge will involve serialization/deserialization) ? AFAIK in the grid there are still references to hibernate proxies in this way and I'm not sure it will work when using a cache which involves serialization.

You are right, definitely not

**Comment by larshelge** [comment]
@d-bernat you think you could add a description for what is changing in this patch? If I read it correctly, are we adding another point of caching grid beyond the one we have in DefaultAnalyticsService?

Instead of including the cache get/put logic inside a method over many lines I think the `analyticsCache.getOrFetch` pattern is more elegant and readable.

**Comment by d-bernat** [comment]
> @d-bernat you think you could add a description for what is changing in this patch? If I read it correctly, are we adding another point of caching grid beyond the one we have in DefaultAnalyticsService?
> 
> Instead of including the cache get/put logic inside a method over many lines I think the `analyticsCache.getOrFetch` pattern is more elegant and readable.

Agree, the pattern will be implemented

Description:
in the class DataAgregator there is critical method getAggregatedDataValueGrid. It is kind of factory. In many steps the grid object is created and filled up with data. In the mid of this process the incomplete Grid object is cached. In the next steps is this object modified (cache has just reference on grid) and finally returned. To avoid caching of incomplete grid object the get operation is now in the very beginning of this factory pipeline and put is in the end of it.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handler/DataHandler.java [hunk 1]
```diff
@@ -985,6 +994,24 @@ private void handleEmptyDimensionItemPermutations( List<List<DimensionItem>> dim
         return result;
     }
 
+    private Grid getOrFetchGrid( final DataQueryParams params, final Function<DataQueryParams, Grid> function )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handler/DataHandler.java:997]: I think this method is already in `AnalyticsCache`#`getOrFetch` ? So we don't need to replicate it here?

Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handler/DataHandler.java:997]: ups, just tested something and forget to remove, sorry. done

Comment (d-bernat) [reply dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handler/DataHandler.java]: sorry, just tested something and forget to remove, it is done.

File: dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handler/DataHandler.java [hunk 2]
```diff
@@ -370,7 +376,7 @@ void addProgramDataElementAttributeIndicatorValues( DataQueryParams params, Grid
             EventQueryParams eventQueryParams = new EventQueryParams.Builder( fromDataQueryParams( dataSourceParams ) )
                 .withSkipMeta( true ).build();
 
-            Grid eventGrid = eventAnalyticsService.getAggregatedEventData( eventQueryParams );
+            Grid eventGrid = eventAnalyticsService.getAggregatedEventData( eventQueryParams, false );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/data/handler/DataHandler.java:379]: Could you instead embed the `tryCache` param part of the `EventQueryParams` object? That is meant to encapsulate parameters. Can be added below the `withSkipMetadata` line.

We could also name the field `skipCacheLookup` with default value of false. I think false is the most sensible value and avoids a breaking change.

**Comment by larshelge** [comment]
@d-bernat this is a temporary fix. I think we could wait for the real fix, as this fix will have a performance impact, and we might not want that plus the additional complexity of introducing, backporting and reverting a temporary fix.

**Pull request closed without merging**